packages/common/states.ts

```diff
@@ -110,15 +52,31 @@ const stateAbbreviations: Record<string, State> = {
   'WV': 'West Virginia',
   'WI': 'Wisconsin',
   'WY': 'Wyoming',
-}
+} as const
+
+type KeysOf <R> = R extends Record<infer K, unknown> ? K : never
+export type StateAbbreviation = KeysOf<typeof stateAbbreviations>
+
+export const allStates = Object.values(stateAbbreviations)
+
+export type State = (typeof allStates)[number]
+// We can't use set here because it seems Google Translator deplays the DOM loading and therefore the resulting set will be empty
+export const isState = (x: string): x is State => allStates.some(state => state === x)
+export type StateField = {state: State}
+export type ExtendsState<T extends State> = T extends State ? T : never
 
 const caseInsensitiveStates: Record<string, State> = Object.fromEntries(
   Object.values(stateAbbreviations)
     .map(state => [state.toUpperCase(), state])
 )
-  
 
 export const getState = (key: string): State | undefined => {
   const upperKey = key.toUpperCase()
-  return stateAbbreviations[upperKey] ?? caseInsensitiveStates[upperKey] ?? undefined
+  return stateAbbreviations[upperKey as StateAbbreviation] ?? caseInsensitiveStates[upperKey] ?? undefined
+}
+
+export const getStateAbbr = (value: State): StateAbbreviation | undefined => {
+  return (Object.keys(stateAbbreviations) as StateAbbreviation[]).find(
```

Comment:
Is ` as StateAbbreviation[]` necessary anymore?

Reply:
Unfortunately it is, Object.keys will still infer `type == string[]`

Reply:
Crazy!